lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -2331,6 +2331,90 @@ static LogicalResult checkConnectFlow(Operation *connect) {
   return success();
 }
 
+namespace {
+enum class ConstLeafKind : uint8_t { None = 0, Normal = 1 << 0, Flip = 1 << 1 };
+} // end anonymous namespace
+
+// NOLINTBEGIN(misc-no-recursion)
+/// Checks if the type has any 'const' leaf elements that are normal or flip
+static ConstLeafKind typeConstLeafKinds(FIRRTLBaseType type,
+                                        bool outerTypeIsConst = false,
+                                        bool isFlip = false) {
+  auto typeIsConst = outerTypeIsConst || type.isConst();
+
+  if (typeIsConst && type.isPassive())
+    return isFlip ? ConstLeafKind::Flip : ConstLeafKind::Normal;
+
+  if (auto bundleType = type.dyn_cast<BundleType>()) {
+    auto kinds = ConstLeafKind::None;
+    for (auto &element : bundleType.getElements()) {
+      auto elementKinds = typeConstLeafKinds(element.type, typeIsConst,
+                                             isFlip ^ element.isFlip);
+      kinds = static_cast<ConstLeafKind>(static_cast<uint8_t>(kinds) |
+                                         static_cast<uint8_t>(elementKinds));
+    }
+    return kinds;
+  }
+
+  if (auto vectorType = type.dyn_cast<FVectorType>())
+    return typeConstLeafKinds(vectorType.getElementType(), typeIsConst, isFlip);
+
+  if (typeIsConst)
+    return isFlip ? ConstLeafKind::Normal : ConstLeafKind::Flip;
+  return ConstLeafKind::None;
+}
```

Comment:
This part looks backwards:
`return isFlip ? ConstLeafKind::Normal : ConstLeafKind::Flip;`

I don't love the use of enum classes here, but I don't think that you have to track both pieces of information.  If you pass in `isFlip=true` when examining the source type, you only have to track if there are "normal const leaf elements", which means this could return a bool.

```c++
static bool isConstFieldDriven(FIRRTLBaseType type,
                                        bool isFlip = false,
                                        bool outerTypeIsConst = false) {
  auto typeIsConst = outerTypeIsConst || type.isConst();

  if (typeIsConst && type.isPassive())
    return !isFlip;

  if (auto bundleType = dyn_cast<BundleType>(type)) {
    for (auto &element : bundleType.getElements()) {
      if (typeConstLeafKinds(element.type, typeIsConst, isFlip ^ element.isFlip))
        return true;
    return false;
  }

  if (auto vectorType = type.dyn_cast<FVectorType>())
    return typeConstLeafKinds(vectorType.getElementType(), typeIsConst, isFlip);

  if (typeIsConst)
    return !isFlip;
  return false;
}
```

And the caller could be written:
```c++
  // Check destination if it contains 'const' leaves.
  if (destType.containsConst() && isConstFieldDriven(destType))
    if (failed(checkConstConditionality(dest, destType)))
      return failure();

  // Check source if it contains 'const' 'flip' leaves
  if (srcType.containsConst() && isConstFieldDriven(srcType, /*isFlip=*/true))
    if (failed(checkConstConditionality(src, srcType)))
      return failure();
```

Reply:
Good point, the enums are especially ugly without adding the bitwise operator overloads, and I guess we don't really need them. Updated to check one type of flippedness.